packages/grafana-ui/src/components/Card/Card.tsx

```diff
@@ -21,83 +18,181 @@ export interface Props extends Omit<CardContainerProps, 'disableEvents' | 'disab
 }
 
 export interface CardInterface extends FC<Props> {
+  Heading: FC<ChildProps>;
   Tags: typeof Tags;
   Figure: typeof Figure;
   Meta: typeof Meta;
   Actions: typeof Actions;
   SecondaryActions: typeof SecondaryActions;
+  Description: typeof Description;
 }
+const validComponents = ['Heading', 'Figure', 'Meta', 'Description', 'Tags', 'Actions', 'SecondaryActions'];
 
 /**
  * Generic card component
  *
  * @public
  */
-export const Card: CardInterface = ({ heading, description, disabled, href, onClick, children, ...htmlProps }) => {
-  const theme = useTheme2();
-  const styles = getCardStyles(theme);
-  const [tags, figure, meta, actions, secondaryActions] = ['Tags', 'Figure', 'Meta', 'Actions', 'SecondaryActions'].map(
-    (item) => {
-      const found = React.Children.toArray(children as React.ReactElement[]).find((child) => {
-        return React.isValidElement(child) && child?.type && (child.type as any).displayName === item;
-      });
+export const Card: CardInterface = ({ disabled, href, onClick, children, ...htmlProps }) => {
+  const styles = useStyles2(getCardStyles);
+  const onCardClick = onClick && !disabled ? onClick : undefined;
+  const subComponents = useMemo(
+    () =>
+      React.Children.map(children, (c) => {
+        const componentType = React.isValidElement(c) && (c.type as any).displayName;
+        if (!componentType) {
+          return c;
+        }
 
-      if (found && React.isValidElement(found)) {
-        return React.cloneElement(found, { disabled, styles, ...found.props });
-      }
-      return found;
-    }
+        if (validComponents.includes(componentType)) {
+          if (componentType === 'Heading') {
+            return React.cloneElement(c, { disabled, styles, href, onClick: onCardClick, ...c.props });
+          } else {
+            return React.cloneElement(c, { disabled, styles, ...c.props });
+          }
+        }
+
+        return c;
+      }),
+    [children, disabled, href, styles, onCardClick]
+  );
```

Comment:
If we're taking the time to refactor this component, I would like to take the chance to maybe take a refactor our approach here (which was present in the previous code).

This style of fiddling with child props, especially based on looking at the displayName seems off too me. It's not composable and not intuitive as to what'll happen I think.

For onClick and Heading, it looks like this is a change in behaviour? Previously `<Card onClick={} />` would be attached the the root, but now its put on just the title? Not knowing the motivations behind this, I think i would suggest to preserve the previous behaviour, and consumers can explicitly attach onClick to `<Card.Heading />` if they want that?

I think then each of the child components could then just call the styles function themselves to get them, and maybe `disabled` could  be handled with a css attribute selector?

For sharing values with children like this, where absolutely necessary I think it's better to use Context to send the values down rather than `.cloneElement`

Reply:
Ah, yeah, the reason for putting onClick/href on the heading vs the card container itself is pretty unintuitive - should've made a comment. The problem is that, if `href` or `onClick` is on the wrapper component itself, certain screen readers won't read out interactive components nested within that wrapper, which is of course the case a lot of the time with `Card` components. You could just put the `href` or `onClick` on the heading, but then you lose the whole card itself being clickable, which is a shame. The solution which I shamelessly stole from [here](https://inclusive-components.design/cards/) is to use pseudo-elements and some CSS trickery.

I don't mind the compound component pattern so much, but I do agree the implementation here could be better. Good point about using context